File: worldmodel.py
import entities
import pygame
import ordered_list
import actions
import occ_grid
import point
import save_load

class WorldModel:
   def __init__(self, num_rows, num_cols, background):
      self.background = occ_grid.Grid(num_cols, num_rows, background)
      self.num_rows = num_rows
      self.num_cols = num_cols
      self.occupancy = occ_grid.Grid(num_cols, num_rows, None)
      self.entities = []
      self.action_queue = ordered_list.OrderedList()

   def within_bounds(self, pt):
      return (pt.x >= 0 and pt.x < self.num_cols and
         pt.y >= 0 and pt.y < self.num_rows)

   def is_occupied(self, pt):
      return (self.within_bounds(pt) and
         occ_grid.get_cell(self.occupancy, pt) != None)

   def find_nearest(self, pt, type):
      oftype = [(e, distance_sq(pt, entities.get_position(e)))
         for e in self.entities if isinstance(e, type)]

      return nearest_entity(oftype)

   def add_entity(self, entity):
      pt = entities.get_position(entity)
      if self.within_bounds(pt):
         old_entity = occ_grid.Grid.get_cell(self.occupancy, pt)
         if old_entity != None:
            entities.clear_pending_actions(old_entity)
         occ_grid.Grid.set_cell(self.occupancy, pt, entity)
         self.entities.append(entity)

   def move_entity(self, entity, pt):
      tiles = []
      if self.within_bounds(pt):
         old_pt = entities.get_position(entity)
         occ_grid.set_cell(self.occupancy, old_pt, None)
         tiles.append(old_pt)
         occ_grid.set_cell(self.occupancy, pt, entity)
         tiles.append(pt)
         entities.set_position(entity, pt)

      return tiles

   def remove_entity(self, entity):
      self.remove_entity_at(entities.get_position(entity))

   def remove_entity_at(self, pt):
      if (self.within_bounds(pt) and
         occ_grid.get_cell(self.occupancy, pt) != None):
         entity = occ_grid.get_cell(elf.occupancy, pt)
         entities.set_position(entity, point.Point(-1, -1))
         self.entities.remove(entity)
         occ_grid.set_cell(self.occupancy, pt, None)

   def schedule_action(self, action, time):
      self.action_queue.insert(action, time)

   def unschedule_action(self, action):
      self.action_queue.remove(action)

   def update_on_time(self, ticks):
      tiles = []

      next = self.action_queue.head()
      while next and next.ord < ticks:
         self.action_queue.pop()
         tiles.extend(next.item(ticks))  # invoke action function
         next = self.action_queue.head()

      return tiles

   def get_background_image(self, pt):
      if self.within_bounds(pt):
         return entities.get_image(occ_grid.get_cell(self.background, pt))


   def get_background(self, pt):
      if self.within_bounds(pt):
         return occ_grid.get_cell(self.background, pt)


   def set_background(self, pt, bgnd):
      if self.within_bounds(pt):
         occ_grid.set_cell(self.background, pt, bgnd)


   def get_tile_occupant(self, pt):
      if self.within_bounds(pt):
         return occ_grid.get_cell(self.occupancy, pt)

   def get_entities(self):
      return self.entities

   """#save_load.py
   def add_background(self, properties, i_store):
      BGND_NUM_PROPERTIES = 4
      BGND_NAME = 1
      BGND_COL = 2
      BGND_ROW = 3
      if len(properties) >= BGND_NUM_PROPERTIES:
         pt = point.Point(int(properties[BGND_COL]), int(properties[BGND_ROW]))
         name = properties[BGND_NAME]
         self.set_background(pt,
            entities.Background(name, image_store.get_images(i_store, name)))

   def load_add_entity(self, properties, i_store, run):
      new_entity = save_load.create_from_properties(properties, i_store)
      if new_entity:
         self.add_entity(new_entity)
         if run:
            self.schedule_entity(new_entity, i_store)

   def schedule_entity(self, entity, i_store):
      if isinstance(entity, entities.MinerNotFull):
         self.schedule_miner(entity, 0, i_store)
      elif isinstance(entity, entities.Vein):
         self.schedule_vein(entity, 0, i_store)
      elif isinstance(entity, entities.Ore):
         self.schedule_ore(entity, 0, i_store) """
        
   #ACTIONS.PY

   

#def within_bounds(world, pt):
#   return (pt.x >= 0 and pt.x < world.num_cols and
#      pt.y >= 0 and pt.y < world.num_rows)


#def is_occupied(world, pt):
#   return (within_bounds(world, pt) and
#      occ_grid.get_cell(world.occupancy, pt) != None)


def nearest_entity(entity_dists):
   if len(entity_dists) > 0:
      pair = entity_dists[0]
      for other in entity_dists:
         if other[1] < pair[1]:
            pair = other
      nearest = pair[0]
   else:
      nearest = None

   return nearest


def distance_sq(p1, p2):
   return (p1.x - p2.x)**2 + (p1.y - p2.y)**2


"""def find_nearest(world, pt, type):
   oftype = [(e, distance_sq(pt, entities.get_positio(e)))
      for e in world.entities if isinstance(e, type)]n

   return nearest_entity(oftype)


def add_entity(world, entity):
   pt = entities.get_position(entity)
   if within_bounds(world, pt):
      old_entity = occ_grid.get_cell(world.occupancy, pt)
      if old_entity != None:
         entities.clear_pending_actions(old_entity)
      occ_grid.set_cell(world.occupancy, pt, entity)
      world.entities.append(entity)


def move_entity(world, entity, pt):
   tiles = []
   if within_bounds(world, pt):
      old_pt = entities.get_position(entity)
      occ_grid.set_cell(world.occupancy, old_pt, None)
      tiles.append(old_pt)
      occ_grid.set_cell(world.occupancy, pt, entity)
      tiles.append(pt)
      entities.set_position(entity, pt)

   return tiles


def remove_entity(world, entity):
   remove_entity_at(world, entities.get_position(entity))


def remove_entity_at(world, pt):
   if (within_bounds(world, pt) and
      occ_grid.get_cell(world.occupancy, pt) != None):
      entity = occ_grid.get_cell(world.occupancy, pt)
      entities.set_position(entity, point.Point(-1, -1))
      world.entities.remove(entity)
      occ_grid.set_cell(world.occupancy, pt, None)


def schedule_action(world, action, time):
   world.action_queue.insert(action, time)


def unschedule_action(world, action):
   world.action_queue.remove(action)


def update_on_time(world, ticks):
   tiles = []

   next = world.action_queue.head()
   while next and next.ord < ticks:
      world.action_queue.pop()
      tiles.extend(next.item(ticks))  # invoke action function
      next = world.action_queue.head()

   return tiles


def get_background_image(world, pt):
   if within_bounds(world, pt):
      return entities.get_image(occ_grid.get_cell(world.background, pt))


def get_background(world, pt):
   if within_bounds(world, pt):
      return occ_grid.get_cell(world.background, pt)


def set_background(world, pt, bgnd):
   if within_bounds(world, pt):
      occ_grid.set_cell(world.background, pt, bgnd)


def get_tile_occupant(world, pt):
   if within_bounds(world, pt):
      return occ_grid.get_cell(world.occupancy, pt)


def get_entities(world):
   return world.entities"""

File: WorldModel.java
import java.util.*;
import java.util.function.*;

import processing.core.*;

public class WorldModel extends PApplet
{
	private Background[][] background;
	public final int num_rows;
	public final int num_cols;
	private Location[][] occupancy;
	private List<Location> entities;
	private OrderedList action_queue;


	public WorldModel(int num_rows, int num_cols, Background background)
	{
		this.background = new Background[num_rows][num_cols];
		this.num_rows = num_rows;
		this.num_cols = num_cols;
		this.occupancy = new Location[num_rows][num_cols];
		this.entities = new ArrayList<Location>();
		this.action_queue = new OrderedList();
		for (int x = 0; x < num_cols; x++)
		{
			for (int y = 0; y < num_rows; y++)
			{
				this.background[y][x] = background;
			}
		}
	}

	public boolean within_bounds(Point pt)
	{
		return pt.x >= 0 && pt.x < num_cols && pt.y >= 0 && pt.y < num_rows;
	}

	public boolean is_occupied(Point pt)
	{
		return within_bounds(pt) && (occupancy[pt.y][pt.x] != null);
	}

	public Location find_nearest(Point pt, Class type)
	{
		List<Location> oftype = new ArrayList<Location>();
		for (Location e : entities)
		{
			if (type.isInstance(e))
			{
				oftype.add(e);
			}
		}
		
      	return nearest_entity(oftype, pt);
	}

	public void add_entity(Location entity)
	{
		Point pt = entity.get_position();
      	if (within_bounds(pt))
      	{
        	Job old_entity = (Job) occupancy[pt.y][pt.x];
        	if (old_entity != null)
        	{
            	old_entity.clear_pending_actions();
        	}
        	occupancy[pt.y][pt.x] = entity;
        	entities.add(entity);
      	}
	}

	public List<Point> move_entity(Location entity, Point pt)
	{
		List<Point> tiles = new ArrayList<Point>();

		if (within_bounds(pt))
		{
         	Point old_pt = entity.get_position();
         	occupancy[old_pt.y][old_pt.x] = null;
         	tiles.add(old_pt);
         	occupancy[pt.y][pt.x] = entity;
         	tiles.add(pt);
         	entity.set_position(pt);
		}

      return tiles;
	}

	public void remove_entity(Location entity)
	{
		remove_entity_at(entity.get_position());
	}

	public void remove_entity_at(Point pt)
	{
		if (within_bounds(pt) && occupancy[pt.y][pt.x] != null)
		{
         	Location entity = occupancy[pt.y][pt.x];
         	Point set_point = new Point(-1, -1);
         	entity.set_position(set_point);
         	entities.remove(entity);
         	occupancy[pt.y][pt.x] = null;
		}
	}

	public void schedule_action(LongConsumer action, long time)
    {
    	this.action_queue.insert(action, time);
    }

    public void unschedule_action(LongConsumer action)
    {
    	this.action_queue.remove(action);
    }

    public void update_on_time(long ticks)
    {
    	List<Rectangle> tiles = new ArrayList<Rectangle>();

      	ListItem next = this.action_queue.head();
      	while (next != null && next.get_ord() < ticks)
        {
        	this.action_queue.pop();
         	//tiles.add(next.get_item().accept(ticks));
         	next = this.action_queue.head();
        }

      	//return tiles;
    }

	public PImage get_background_image(Point pt)
	{
		if (within_bounds(pt))
		{
    		return background[pt.y][pt.x].get_image();
		}
		return null;
	}
	
	public Background get_background(Point pt)
	{
		if (within_bounds(pt))
		{
    		return background[pt.y][pt.x];
    	}
		return null;
	}

	public void set_background(Point pt, Background bgnd)
	{
		if (within_bounds(pt))
		{
         	background[pt.y][pt.x] = bgnd;
		}
	}

	public Location get_tile_occupant(Point pt)
	{
		if (within_bounds(pt))
		{
         	return occupancy[pt.y][pt.x];
		}
		return null;
	}

	public List<Location> get_entities()
	{
		return entities;
	}
	
	private Location nearest_entity(List<Location> entity_list, Point pt)
    {
        if(entity_list.size() > 0)
        {
        	Location closest = entity_list.get(0);
            for(Location entity : entity_list)
            {
            	if(distance_sq(pt, entity.get_position()) < distance_sq(pt, closest.get_position()))
            	{
            		closest = entity;
            	}
            }
            return closest;
        }
        return null;
    }
	
	private double distance_sq(Point p1, Point p2)
	{
		return (p1.x - p2.x)*(p1.x - p2.x) + (p1.y - p2.y)*(p1.y - p2.y);
	}
}